File: STRINGS/876-hand-of-straights/hand-of-straights.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        # Check if the total number of cards is divisible by groupSize
        if len(hand) % groupSize != 0:
            return False

        # Count the frequency of each card
        card_count = Counter(hand)

        # Iterate over the cards in sorted order
        for card in sorted(card_count):
            if card_count[card] > 0:  # If there are any of this card left to place in groups
                count = card_count[card]
                for i in range(groupSize):
                    if card_count[card + i] < count:
                        return False
                    card_count[card + i] -= count

        return True

File: STRINGS/876-hand-of-straights/hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        map<int, int> mp;
        int n = size(hand);
        if (n % groupSize) {
            return false;
        }
        for (int& handNum : hand) {
            mp[handNum]++;
        }
        while (!mp.empty()) {
            int curr = mp.begin()->first;
            for (int i = 0; i < groupSize; ++i) {
                if (mp[curr + i] == 0) {
                    return false;
                }
                mp[curr + i]--;
                if (mp[curr + i] < 1) {
                    mp.erase(curr + i);
                }
            }
        }
        return true;
    }
};